x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/TypeResolutionUtils.java

```diff
@@ -0,0 +1,129 @@
+        return typeMustBe(e, DataType::isNumeric, operationName, paramOrd, "numeric");
+    }
+
+    public static TypeResolution typeMustBeString(Expression e, String operationName, ParamOrdinal paramOrd) {
+        return typeMustBe(e, DataType::isString, operationName, paramOrd, "string");
+    }
+
+    public static TypeResolution typeMustBeDate(Expression e, String operationName, ParamOrdinal paramOrd) {
+        return typeMustBe(e, DataType::isDateBased, operationName, paramOrd, "date", "datetime");
+    }
+
+    public static TypeResolution typeMustBeNumericOrDate(Expression e, String operationName, ParamOrdinal paramOrd) {
+        return typeMustBe(e, dt -> dt.isNumeric() || dt.isDateBased(), operationName, paramOrd, "date", "datetime", "numeric");
+    }
+
+    public static TypeResolution typeMustBeExact(Expression e, String message) {
+        if (e instanceof FieldAttribute) {
+            Tuple<Boolean, String> hasExact = ((FieldAttribute) e).hasExact();
+            if (hasExact.v1() == Boolean.FALSE) {
+                return new TypeResolution(format(null, message, e.dataType().typeName, hasExact.v2()));
+            }
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    public static TypeResolution typeMustBeExact(Expression e, String operationName, ParamOrdinal paramOrd) {
+        if (e instanceof FieldAttribute) {
+            Tuple<Boolean, String> hasExact = ((FieldAttribute) e).hasExact();
+            if (hasExact.v1() == Boolean.FALSE) {
+                return new TypeResolution(format(null, "[{}] cannot operate on {}field of data type [{}]: {}",
+                    operationName,
+                    paramOrd == null || paramOrd == ParamOrdinal.DEFAULT ?
+                        "" : paramOrd.name().toLowerCase(Locale.ROOT) + " argument ",
+                    e.dataType().typeName, hasExact.v2()));
+            }
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    public static TypeResolution typeMustBeStringAndExact(Expression e, String operationName, ParamOrdinal paramOrd) {
+        TypeResolution resolution = typeMustBe(e, DataType::isString, operationName, paramOrd, "string");
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        return typeMustBeExact(e, operationName, paramOrd);
+    }
+
+    public static TypeResolution expressionMustBeConstant(Expression e, String operationName, ParamOrdinal paramOrd) {
+        if (!e.foldable()) {
+            return new TypeResolution(format(null, "{}argument of [{}] must be a constant, received [{}]",
+                paramOrd == null || paramOrd == ParamOrdinal.DEFAULT ? "" : paramOrd.name().toLowerCase(Locale.ROOT) + " ",
+                operationName,
+                Expressions.name(e)));
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    public static TypeResolution expressionMustBeTableColumn(Expression e, String operationName, ParamOrdinal paramOrd) {
```

Comment:
constant/table column is a bit misleading - I would just use `isFoldable`/`notFoldable` since it's a terminology already used through-out the code.